src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -11,13 +11,13 @@ namespace System.Net.Sockets
     public partial class SocketAsyncEventArgs : EventArgs, IDisposable
     {
         // AcceptSocket property variables.
-        internal Socket _acceptSocket;
+        private Socket _acceptSocket;
         private Socket _connectSocket;
 
-        // Buffer,Offset,Count property variables.
-        internal byte[] _buffer;
-        internal int _count;
-        internal int _offset;
+        // Single buffer.
+        private Memory<byte> _buffer;
+        private int _offset;
```

Comment:
Do we actually need offset/count anymore?  Why isn't it just _buffer?

Reply:
If someone does:
```C#
e.SetBuffer(new byte[10], 0, 10);
e.SetBuffer(5, 1);
e.SetBuffer(0, 10);
```
we need that to continue working, but if the implementation slices the buffer and overwrites the original with the slice, then we've lost the ability to change the offset and count.

Reply:
Can't we just do TryGetArray?

Reply:
> Can't we just do TryGetArray?

That would then allow:
```C#
Memory<T> memory = ...; // someone's handed in just the portion that should be used
e.SetBuffer(memory);
e.SetBuffer(0, 1000); // oops
```
as we can't tell the difference at that point whether we're allowed to use the relevant portion of the array or not.

Reply:
More generally, what happens when I call .Buffer, .Offset, or .Count on (a) a byte[]-backed Memory<byte>?  A non-byte[]-backed Memory<byte>?

What happens when I call SetBuffer(offset, count) in these two cases?

Reply:
> More generally, what happens when I call .Buffer, .Offset, or .Count on (a) a byte[]-backed Memory?

.Buffer gives you back the underlying array
.Offset gives you back the passed in offset
.Count gives you back the passed in count

But that makes me realize there's a bug here.  .Offset should give you back the offset from the array segment + _offset, and the same for arraySegment.Count + _count.  I'll fix that.

>  A non-byte[]-backed Memory?

.Buffer gives you null
.Offset gives you 0
.Length gives you 0

But once I fix the above, Length will give you memory.Length.

> What happens when I call SetBuffer(offset, count) in these two cases?

It updates _offset and _count accordingly.  So you can pass in a memory and then use SetBuffer(offset, count) to jump around in it, as you can with arrays.

Reply:
> But that makes me realize there's a bug here. 

Actually, hmm, now I'm not sure what we should do in this case.  Thinking...

Reply:
The problem is this.  If someone does:
```C#
e.SetBuffer(array, x, y);
byte[] resultingArray = e.Buffer;
int resultingOffset = e.Offset;
int resultingCount = e.Count;
```
resultingArray should match array, resultingOffset should match x, and resultingCount should match y.  That's how it works today.  And with the code in this PR, if you do:
```C#
e.SetBuffer((Memory<byte>)array);
e.SetBuffer(x, y);
int resultingOffset = e.Offset;
int resultingCount = e.Count;
```
again resultingOffset and resultingCount will match x and y respectively.  But, if you do:
```C#
e.SetBuffer(new Memory<byte>(array, 3, array.Length - 3));
e.SetBuffer(1, 1);
byte[] resultingArray = array;
int resultingOffset = e.Offset;
int resultingCount = e.Count;
```
resultingOffset will be 1 and resultingCount will be 1, but if you used those to index into array, you'd get the wrong piece, i.e. you'd get the element at array[1] rather than the element at array[3+1].  I could change the implementation so that instead of doing:
```C#
public int Offset => _offset;
public int Count => _count;
```
it does:
```C#
public int Offset => _memory.TryGetArray(out var s) ? _offset + s.Offset : _offset;
public int Count => _count;
```
and that would make the previous example:
```C#
e.SetBuffer(new Memory<byte>(array, 3, array.Length - 3));
e.SetBuffer(1, 1);
byte[] resultingArray = array;
int resultingOffset = e.Offset;
int resultingCount = e.Count;
```
yield resultingOffset as 4, which is correct for indexing into the array, but it also means that the values passed in to SetBuffer(int, int) don't necessarily roundtrip, in the case where `SetBuffer(Memory<byte>)` is used.  That seems like the least bad option.  Thoughts?

Reply:
My suggestion would be 

(a) Don't support Offset/Length for any Memory<T> (byte[] or not).  It's easy enough to slice yourself, and having an added slicing mechanism is confusing.
(b) Don't support Buffer for any Memory<T> (byte[] or not).  Just return null.

In addition, I wonder if we should have a field like MemoryBuffer to expose the Memory<byte> instead of having GetBuffer().  It's a little weird that we have both .Buffer and .GetBuffer().

Reply:
Also

(c) Don't support SetBuffer(offset, count) on any Memory\<byte\>, this should throw (this is in line with my comment above re not supporting Offset/Length in this case)

Reply:
> My suggestion would be

So track in a bool whether `SetBuffer(byte[],int,int)` or `SetBuffer(Memory<byte>)` was used, and if the latter was used, throw from `SetBuffer(int,int)` and return null/0 from Buffer/Offset/Count?

> In addition, I wonder if we should have a field like MemoryBuffer to expose the Memory instead of having GetBuffer(). It's a little weird that we have both .Buffer and .GetBuffer().

A field?  Or did you just mean change the `GetBuffer()` method to be a `Memory { get; }` property?

Reply:
> So track in a bool

Yeah, or just save off the original byte[] in the non-memory case.  That would avoid a call to TryGetArray.  Either way works.  Saving the original byte[] might be slightly less confusing.

Yeah, I meant property not field.  Something like .Memory or .MemoryBuffer etc.